hollow/src/main/java/com/netflix/hollow/api/producer/SealableWriteState.java

```diff
@@ -25,29 +25,36 @@
  *
  * @author Tim Taylor {@literal<tim@toolbear.io>}
  */
-final class WriteStateImpl implements HollowProducer.WriteState {
+final class SealableWriteState implements HollowProducer.WriteState {
     private final long version;
     private final HollowObjectMapper objectMapper;
     private final HollowProducer.ReadState priorReadState;
+    private volatile boolean sealed = false;
 
-    protected WriteStateImpl(long version, HollowObjectMapper objectMapper, HollowProducer.ReadState priorReadState) {
+    SealableWriteState(long version, HollowObjectMapper objectMapper, HollowProducer.ReadState priorReadState) {
         this.version = version;
         this.objectMapper = objectMapper;
         this.priorReadState = priorReadState;
     }
 
     @Override
-    public int add(Object o) {
+    public int add(Object o) throws SealedWriteStateException {
+        if (sealed)
+            throw new SealedWriteStateException(this);
         return objectMapper.add(o);
     }
 
     @Override
-    public HollowObjectMapper getObjectMapper() {
+    public HollowObjectMapper getObjectMapper() throws SealedWriteStateException {
+        if (sealed)
+            throw new SealedWriteStateException(this);
         return objectMapper;
     }
 
     @Override
-    public HollowWriteStateEngine getStateEngine() {
+    public HollowWriteStateEngine getStateEngine() throws SealedWriteStateException {
+        if (sealed)
+            throw new SealedWriteStateException(this);
```

Comment:
`HollowIncrementalProducer` relies on `getStateEngine()`. I did a cursory look at it seems that all uses are before sealing, but I'd want to review it more thoroughly and try it out before releasing this.